Application/EdFi.Ods.Features/OpenApiMetadata/Providers/OpenApiV3UpconversionProvider.cs

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using EdFi.Common.Extensions;
+using EdFi.Ods.Api.Configuration;
+using EdFi.Ods.Api.Extensions;
+using EdFi.Ods.Common.Configuration;
+using EdFi.Ods.Common.Configuration.Sections;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+using Microsoft.OpenApi;
+using Microsoft.OpenApi.Extensions;
+using Microsoft.OpenApi.Models;
+using Microsoft.OpenApi.Readers;
+
+namespace EdFi.Ods.Features.OpenApiMetadata.Providers;
+
+/// <summary>
+/// Provides a mechanism for upconverting an OpenAPI v2 json document to an OpenAPI v3 document while also transforming the servers configuration.
+/// </summary>
+public class OpenApiV3UpconversionProvider : IOpenApiUpconversionProvider
+{
+    private readonly ApiSettings _apiSettings;
+    private readonly ReverseProxySettings _reverseProxySettings;
+    private readonly IEdFiAdminRawOdsInstanceConfigurationDataProvider _IEdFiAdminRawOdsInstanceConfigurationDataProvider;
+    private readonly IHttpContextAccessor _httpContextAccessor;
+    private readonly IOptionsMonitor<TenantsSection> _tenantsConfigurationOptions;
+
+    public OpenApiV3UpconversionProvider(ApiSettings apiSettings,
+        IEdFiAdminRawOdsInstanceConfigurationDataProvider IEdFiAdminRawOdsInstanceConfigurationDataProvider,
+        IHttpContextAccessor httpContextAccessor,
+        IOptionsMonitor<TenantsSection> tenantsConfigurationOptions)
+    {
+        _apiSettings = apiSettings;
+        _reverseProxySettings = apiSettings.GetReverseProxySettings();
+        _IEdFiAdminRawOdsInstanceConfigurationDataProvider = IEdFiAdminRawOdsInstanceConfigurationDataProvider;
+        _httpContextAccessor = httpContextAccessor;
+        _tenantsConfigurationOptions = tenantsConfigurationOptions;
+    }
+
+    public string GetUpconvertedOpenApiJson(string openApiJson)
+    {
+        var openApiDocument = new OpenApiStringReader().Read(openApiJson, out _);
+        PopulateServersConfiguration(ref openApiDocument);
+        return openApiDocument.SerializeAsJson(OpenApiSpecVersion.OpenApi3_0);
+    }
+
+    private void PopulateServersConfiguration(ref OpenApiDocument openApiDocument)
+    {
+        openApiDocument.Servers.Clear();

```

Comment:
This method only gets called once, when this provider (which is configured as single instance) is constructed. So whichever servers are fetched the first time this is instantiated will be the same for all other tenants.